airflow/models/taskinstance.py

```diff
@@ -1446,18 +1454,21 @@ def handle_failure(self, error, test_mode=None, context=None, force_fail=False,
                 self.log.error('Failed to send email to: %s', task.email)
                 self.log.exception(exec2)
 
-        # Handling callbacks pessimistically
-        if callback:
-            try:
-                callback(context)
-            except Exception as exec3:  # pylint: disable=broad-except
-                self.log.error("Failed at executing callback")
-                self.log.exception(exec3)
-
         if not test_mode:
             session.merge(self)
         session.commit()
 
+    @provide_session
+    def handle_failure_with_callback(
+        self,
+        error,
+        test_mode=None,
+        force_fail: bool = False,
+        session=None,
```

Comment:
Should we add type hints?

Reply:
@turbaszek added type hints for error and test_mode, but leaving out session since I am not sure what's the best way to handle because it's not supposed to be `Optional[Session]` due to the provide_session decorator. But on the other hand we need to set the default value to `None` so the provide_session operator knows when to inject the session.